File: rpc/serialize.js
//TODO(sammccall): handle cyclic structures

serialize = (function() {
  function Transform(test, apply) {
    this.test = test;
    this.apply = apply;
  }

  function Transforms() {
    this.items = [];
    var self = this;
    this.register(function(x) { return typeof x == 'object' && x != null; }, function(obj) {
      var result = {};
      for (var key in obj) {
        if (obj.hasOwnProperty(key)) result[key] = self.apply(obj[key]);
      }
      return result;
    });
    this.register(function(x) { return x instanceof Array; }, function(list) {
      var result = new Array(list.length);
      for (var i = 0; i < list.length; i++) {
        result[i] = self.apply(list[i]);
      }
      return result;
    });
  }

  Transforms.prototype.register = function(test, apply) {
    this.items.push(new Transform(test, apply));
  }

  Transforms.prototype.apply = function(object) {
    for (var i = this.items.length - 1; i >= 0; i--) {
      if (this.items[i].test(object)) return this.items[i].apply(object);
    }
    return object;
  }

  function Serializer() {
    Transforms.call(this);
    this.register(function(x) { return typeof x == 'undefined'; }, function(x) { return null; });
  }
  Serializer.prototype = Object.create(Transforms.prototype);

  function Deserializer() {
    Transforms.call(this);
  }
  Deserializer.prototype = Object.create(Transforms.prototype);

  return {
    Serializer: Serializer,
    Deserializer: Deserializer,
  };
})();

File: rpc/serialize.dart
#library('serialize');

class _Transform {
  final Function test;
  final Function apply;
  _Transform(this.test, this.apply);
}

class Transforms {
  final List<_Transform> items;
  Transforms() : items = [] {
    register((x) => x is Map, function(map) {
      final result = new Map();
      for (final key in map.getKeys()) result[key] = this.apply(map[key]);
      return result;
    });
    register((x) => x is List, function(list) {
      final result = new List();
      for (final item in list) result.add(this.apply(item));
      return result;
    });
  }

  register(bool test(obj), Object apply(obj)) {
    items.add(new _Transform(test, apply));
  }

  //TODO(sammccall): handle cyclic structures
  Object apply(obj) {
    for (int i = items.length - 1; i >= 0; i--) {
      if (items[i].test(obj)) return items[i].apply(obj);
    }
    return obj;
  }
}

class Serializer extends Transforms {}
class Deserializer extends Transforms {}
